File: oct2024/intervals.js
/**
 * Definition of Interval:
 * class Interval {
 *   constructor(start, end) {
 *     this.start = start;
 *     this.end = end;
 *   }
 * }
 */

// export class Solution {
//   /**
//    * @param intervals: an array of meeting time intervals
//    * @return: if a person could attend all meetings
//    */
var canAttendMeetings = function (intervals) {
  //receive an array of arrays of start and end time
  //return true if they can attend all meetings, else return false
  //ex: [[0, 1], [2, 5], [2, 3]] => return false

  //         -----
  //         --------------
  //-----
  // 0   1.   2.  3.   4.  5

  //iterate through the entire intervals
  //set up pointers to look at the first intervals, and look at the second intervals
  //check if the start time of the second interval < end time
  //return false
  //else return true

  //if no intervals, return true
  if (!intervals) return true;

  for (let i = 0; i < intervals.length; i++) {
    //[2, 5]
    for (let j = i + 1; j < intervals.length; j++) {
      //[2,3]
      const [start1, end1] = intervals[i]; //[2,5]
      const [start2, end2] = intervals[j]; //[2, 3]

      if (start2 < end1 && start1 < end2) {
        console.log(start1, end1, start2, end2);
        //2 < 5 && 2< 3
        return false;
      }
    }
  }
  return true;
};
// }
// console.log(
//   canAttendMeetings([
//     [0, 1],
//     [2, 5],
//     [2, 3],
//   ])
// );

/**
 * Definition of Interval:
 * class Interval {
 *   constructor(start, end) {
 *     this.start = start;
 *     this.end = end;
 *   }
 * }
 */

//   export class Solution {
//     /**
//      * @param intervals: an array of meeting time intervals
//      * @return: the minimum number of conference rooms required
//      */
var minMeetingRooms = function (intervals) {
  //receive an array of arrays
  //return the minRoomsNeeded
  //ex: [[0, 1], [2, 5], [2, 3]]
  //      1      0   1. +    1 = 2

  //initialize a variable to track minRoomsNeeded
  //iterate through the entire intervals
  //keep track of the first interval, keep track of second interval
  //check if the start2 < end1,
  //add conference rooms
  //update minRoomsNeeded to the amount that we're currently seeing
  //return minRoomsNeeded

  let minRoomsNeeded = 0;

  for (let i = 0; i < intervals.length; i++) {
    let additionalRooms = 1;
    for (let j = i + 1; j < intervals.length; j++) {
      const [start1, end1] = intervals[i];
      const [start2, end2] = intervals[j];

      if (start1 < end2 && start2 < end1) {
        additionalRooms++;
      }
    }
    minRoomsNeeded = Math.max(minRoomsNeeded, additionalRooms);
  }

  return minRoomsNeeded;
};

// console.log(
//   minMeetingRooms([
//     [0, 1],
//     [2, 5],
//     [2, 3],
//   ])
// );

/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  //receive an array of arrays
  //return a new array of arrays where overlapping intervals are combined together
  //ex: [[0, 1], [2, 5], [2, 3]] => [[0,1], [2, 5]]

  //         -----
  //         --------------
  //-----
  // 0   1.   2.  3.   4.  5

  //initialize a variable to hold the first interval
  //iterate starting at the first index
  //check if the end time of the second interval < end time of the first interval
  //if so, update end interval to the max value seen
  //else, just add in the interval

  //if overlapinteral is empty, return an array of array

  // if (!intervals) return [[]]
  console.log("intervals before sort:", intervals);
  intervals.sort((a, b) => a[0] - b[0]);
  console.log(intervals);

  let overlapInterval = [intervals[0]]; //[2, 5]

  for (let i = 1; i < intervals.length; i++) {
    const [start1, end1] = intervals[i]; //[2,3]
    const results = overlapInterval[overlapInterval.length - 1]; //[[2,5]]

    //ex: [[0, 1], [2, 5], [2, 3]] => [[0,1], [2, 5]]
    if (results[1] >= start1) {
      //5 < 2
      console.log(results[1]);
      results[1] = Math.max(end1, results[1]);
    } else {
      overlapInterval.push(intervals[i]);
    }
  }
  return overlapInterval;

  //time: O(n)
  //space: O(n)
};

console.log(
  merge([
    [1, 4],
    [0, 1],
  ])
);


function canAttendMeetings(intervals) {
  //receive an array of arrays
  //return true if person can attend all meetings, else false
  //example: [[0, 1], [3, 4]] => true
  //example: [[1, 4], [2, 5]] => false
  
  //edge cases: if the intervals are empty, return true b/c they can make it all meetings
  
  //brute force: 
  //iterate through intervals
  //compare first pair of intervals with second pairs
  //check if the start of second interval <= first pair end time
  //if so, return false
  //after checking everything, return true
  
  if (!intervals) return true
   
  for (let i = 0; i < interals.length; i++) {
      for (let j = i + 1; j < intervals.length ;j++) {
          const [start1, end1] = intervals[i];
          const [start2, end2] = intervals[j];
  
          if (start2 < end1 && end2 < start1) {
              return false
          }
      }
  }
  return true
  
  //time for brute force: O(n);
  //space for brute force: O(n)
  
  //optimized method: do one loop
  //sort the arrays based on start time
  //check if the start time is less than end time
  //return false
  //else checked all the intervals, return true
  
  intervals.sort((a,b) => a[0] - b[0]);
  
  for (let i = 1; i < intervals.length; i++) {
      if (intervals[i][0] < intervals[i - 1][1]) {
          return false
      }
  }
  return true
  
  

File: jan2025/intervals.py
from typing import List

def attend_meeting(intervals: List[List[int]]) -> bool:
    if len(intervals) <=1:
        return True
    
    intervals.sort(key=lambda x: x[0])

    for i in range(len(intervals) - 1): 
        if intervals[i][1] > intervals[i + 1][0]:
            return False
    return True

# print(attend_meeting([(1,3), (2,5)]))

class Interval:
    def __init__(self, start: int, end:int):
        self.start = start
        self.end = end

def min_meeting(intervals: List[List[int]]) -> int:
    #receive a list of integers
    #return min amount of rooms needed
     #ex: [(1,3), (2, 5), (6, 7)] => 2
    
    #keep track of rooms needed
    #sort start times, end times

    #iterate through the list with start and end pointer
    #if end > start, increment rooms, move start pointer
    #else subtract rooms, increment end pointer
    #return min rooms

    minRooms = 0

    start = sorted(intervals.start for interval in intervals)
    end = intervals[i].sort(intervals.end for interval in intervals)

    startPtr, endPtr = 0,0

    while startPtr < len(start) and endPtr < len(end):
        if start[startPtr] < end[endPtr]:
            minRooms += 1
            startPtr += 1
        else:
            minRooms -=1
            endPtr += 1
    return minRooms

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        #receive a list of list of integers
        #return a list where all intervals overlapping merged together

        #ex: [[1,4], [2, 5], [7,8]] => [[1,5], [7,8]]

        #sort the list of integers based on start time
        #create an empty list set with the first list from intervals
        #iterate through the list of arrays starting at the second interval
        #check if end time > start time
        #update the end time of the interval
        #else merge in the interval
        #return new result

        intervals.sort(key = lambda x:x[0])

        result = [intervals[0]]

        for i in range(1, len(intervals)):
            [start, end] = intervals[i]
            [prevStart, prevEnd] = result[-1]
            if prevEnd >= start:
                result[-1][1] = max(prevEnd, end)
            else:
                result.append(intervals[i]) 
        return result
    
    class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        #receive a list of lists
        #return the list where the lists are merged in

        #ex: [[1,3], [7,8]] => [[1,3], [7,8]]

        #sort the lists by start time
        #initialize a result list to with the first interval from intervals
        #iterate through the intervals
        #check if the start time of current interval <= end time
        #if so, update the the end time to max end time
        #else
        #add in the interval b/c no cnflict
        #return results

        intervals.sort(key=lambda x:x[0])

        results = [intervals[0]]

        for i in range(1, len(intervals)):
            currentStart, currentEnd = intervals[i]
            prevStart, prevEnd = results[-1]
            if currentStart <= prevEnd:
                results[-1][1]= max(currentEnd, prevEnd)
            else:
                results.append(intervals[i])
        return results
        

        #insert intervals:
        def insert_interval(intervals, new_interval):
            result = []
            i = 0
            
            while i < len(intervals) and intervals[i][1] < new_interval[0]:
                result.append(intervals[i])
                i += 1

            while i < len(intervals) and intervals[i][0] <= new_interval[1]:
                newInterval[0]= min(newInterval[0], intervals[i][0])
                newInterval[1] = max(newInterval[1], intervals[i][1])
                i += 1
            result.append(new_interval)

            while i < len(interval):
                result.append(intervals[i])
                i += 1
            return result

            class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        #receive a list of lists
        #return a new list inserting in the newInterval
        #ex: [[3,5], [1,3]], [6,8] => [[1,5], [6,8]]

        #sort the intervals based on start time
        #iterate through the list
        #have an i pointer
        #iterate through the intervals
        #check if the interval is less than the new interval start
        #update the start and end time of new interval
        #return the new list

        # result = []
        # i = 0

        # while i < len(intervals) and intervals[i][1] < newInterval[0]:
        #     result.append(intervals[i])
        #     i += 1
        
        # while i < len(intervals) and intervals[i][0] <= newInterval[1]:
        #     newInterval[0] = min(intervals[i][0], newInterval[0])
        #     newInterval[1] = max(intervals[i][1], newInterval[1])
        #     i += 1
        # result.append(newInterval)

        # while i < len(intervals):
        #     result.append(intervals[i])
        #     i += 1

        # return result

        
        intervals.append(newInterval)
        intervals.sort()

        res = [intervals[0]]

        for i in range(1, len(intervals)):
            if intervals[i][0] <= res[-1][1]:
                res[-1][1] = max(res[-1][1], intervals[i][1])
            else:
                res.append(intervals[i])
        return res

        

import heapq

def min_meeting_rooms(interval):
    if not intervals:
        return 0
    
    intervals.sort(key = lambda x:x[0])
    heap = []
    heapq.heappush(heap, intervals[0][1])

    for interval in intervals[1:]:
        if interval[0] >= heap[0]:
            heapq.heappop(heap)
        heapq.heappush(heap, interval[1])
    return len(heap)